xclim/generic.py

```diff
@@ -0,0 +1,265 @@
+        def func(x):
+            return dc.isf(1./t, *x)
+    elif mode in ['min', 'low']:
+        def func(x):
+            return dc.ppf(1./t, *x)
+    else:
+        raise ValueError("mode `{}` should be either 'max' or 'min'".format(mode))
+
+    data = dask.array.apply_along_axis(func, p.get_axis_num('dparams'), p)
+
+    # Create coordinate for the return periods
+    coords = dict(p.coords.items())
+    coords.pop('dparams')
+    coords['return_period'] = t
+
+    # Create dimensions
+    dims = list(p.dims)
+    dims.remove('dparams')
+    dims.insert(0, u'return_period')
+
+    # TODO: add time and time_bnds coordinates (Low will work on this)
+    # time.attrs['climatology'] = 'climatology_bounds'
+    # coords['time'] =
+    # coords['climatology_bounds'] =
+
+    out = xr.DataArray(data=data, coords=coords, dims=dims)
+    out.attrs = p.attrs
+    out.attrs['standard_name'] = '{0} quantiles'.format(dist)
+    out.attrs['long_name'] = '{0} return period values for {1}'.format(dist, getattr(arr, 'standard_name', ''))
+    out.attrs['cell_methods'] = (out.attrs.get('cell_methods', '') + ' dparams: ppf').strip()
+    out.attrs['units'] = arr.attrs.get('units', '')
+    out.attrs['mode'] = mode
+    out.attrs['history'] = out.attrs.get('history', '') + "Compute values corresponding to return periods."
+
+    return out
+
+
+def frequency_analysis(da, mode, t, dist, window=1, freq='YS', **indexer):
+    """Return the value corresponding to a return period.
+
+    Parameters
+    ----------
+    da : xarray.DataArray
+      Input data.
+    t : int or sequence
+      Return period. The period depends on the resolution of the input data. If the input array's resolution is
+      yearly, then the return period is in years.
+    dist : str
+      Name of the univariate distribution, such as beta, expon, genextreme, gamma, gumbel_r, lognorm, norm
+      (see scipy.stats).
+    mode : {'min', 'max'}
+      Whether we are looking for a probability of exceedance (high) or a probability of non-exceedance (low).
+    window : int
+      Averaging window length (days).
+    freq : str
+      Resampling frequency. If None, the frequency is assumed annual and adjusted to the seasons. That is,
+      years will start in December if season is 'DJF'.
+    **indexer : {dim: indexer, }, optional
+      Time attribute and values over which to subset the array. For example, use season='DJF' to select winter values,
+      month=1 to select January, or month=[6,7,8] to select summer months. If not indexer is given, all values are
+      considered.
+
+    Returns
+    -------
+    xarray.DataArray
+      An array of values with a 1/t probability of exceedance or non-exceedance when mode is high or low respectively.
+
+    """
+    # Apply rolling average
+    if window > 1:
+        da = da.rolling(time=window, center=False).mean()
+
+    # Assign default resampling frequency if not provided
+    freq = freq or default_freq(**indexer)
```

Comment:
There is a default freq='YS' in the function parameters so default_freq() would never be used?

Reply:
i.e Should the function def have freq=None ?

Reply:
Unless you pass None. You've put your finger on the tender spot of the PR. Not clear to me what to do here. One option is to set the default frequency based on the period. So for example, it's always 'YS' unless seasons is DJF. The issue with this is that you might want to take the full year, but split the year in July. So, I think we need to expose the `freq` option, the question is how to handle the default. Options: 
- Set freq=None, and then call default_freq to get the default. The problem is that it's not apparent in the signature what the default value is going to be;
- Set freq='YS', and only call default_freq if freq is None;
- Set freq as a positional argument, so users need to pass a value.

Reply:
Now I remember ! But the reason I've chosen this option is that our Indicator class expects the signature to contain freq and its default. If we change that, I need to go back in Indicators and modify the code. Doable, but I was lazy. So let's decide how it should be, and I'll make it work.

Reply:
With the code as is I think many users would not think to explicitly set freq=None in order to get the appropriate default for the DJF season (despite the fact that you wrote it in the docstring. e.g. I missed it and was looking fairly closely )

Reply:
> * Set freq=None, and then call default_freq to get the default. The problem is that it's not apparent in the signature what the default value is going to be;

If changes to the Indicator class are not a major headache .. this would maybe be the best option in my mind.  Could the docstring include a more explicit mention of the defaults by season generated by default_freq()

Reply:
Yes, let's try that.

Reply:
Done.